libraries/Microsoft.Bot.Builder.Azure/CosmosDbStorage.cs

```diff
@@ -230,14 +200,22 @@ public async Task WriteAsync(IDictionary<string, object> changes, CancellationTo
                 if (etag == null || etag == "*")
                 {
                     // if new item or * then insert or replace unconditionaly
-                    await _client.UpsertDocumentAsync(collectionLink, documentChange, disableAutomaticIdGeneration: true).ConfigureAwait(false);
+                    await _client.UpsertDocumentAsync(
+                        _collectionLink,
+                        documentChange,
+                        disableAutomaticIdGeneration: true,
+                        cancellationToken: cancellationToken).ConfigureAwait(false);
                 }
                 else if (etag.Length > 0)
                 {
                     // if we have an etag, do opt. concurrency replace
                     var uri = UriFactory.CreateDocumentUri(_databaseId, _collectionId, documentChange.Id);
                     var ac = new AccessCondition { Condition = etag, Type = AccessConditionType.IfMatch };
-                    await _client.ReplaceDocumentAsync(uri, documentChange, new RequestOptions { AccessCondition = ac }).ConfigureAwait(false);
+                    await _client.ReplaceDocumentAsync(
+                        uri,
+                        documentChange,
+                        new RequestOptions { AccessCondition = ac },
+                        cancellationToken: cancellationToken).ConfigureAwait(false);
                 }
                 else
```

Comment:
just a suggestion... isn't a little bit of an odd structure to leave this to the bottom?
wouldn't it be clearer to just check if (eTag == string.Empty) at the top?